forem/forem
# Runtime class for standardized native features

## What type of PR is this? (check all applicable)


## Description

Another step towards better global management of mobile native features. The new `Runtime` class can be used for checking whether a specific feature is available in the current context. It supports the use of a namespace for more granular checks, which helps deploy logic to the web JS before the native app release is published. This prevents breaking current/older native app releases.

__What's included in the PR:__
- All `isNativeIOS`/`isNativeAndroid` functions spread throughout the codebase now live in a single location
- Fixes #9962 by making use of the native share feature on Android devices
- Replaces the Google Analytics events from Podcasts with Ahoy in the same way as the Video Player (see #9539)


__What's not included in the PR:__
The Strategy pattern for Podcasts ([suggested here](https://github.com/forem/forem/pull/8054#discussion_r430210129)) is not fully implemented, because it requires a more thorough refactor of the Podcast player. It should ideally pick a strategy and reuse the code without caring whether we're working with mobile vs web.

This is an improvement though, because in the past we were doing checks to see if it was Android/iOS/web many times throughout the player. With this refactor we now make use of the Strategy pattern by storing the correct strategy during initialization in `Runtime.podcastMessage` and only if `undefined` we execute the web logic. This reduces from many checks to one `if (Runtime.podcastMessage) { ... } else { ... }`).

## Related Tickets & Documents

#9962 

## QA Instructions, Screenshots, Recordings

The copy to clipboard features in the following locations should continue to work in web & mobile:
- Image upload in Editor
- Connect share invitation link
- Article show page dropdown (copy URL)

The Podcast & video player should work both in web & mobile apps.

I understand checking that the native app features continue to work is tricky. Here's a video of it working but if interested in trying it out locally feel free to ask for help and I'll be happy to walk through the process.

https://share.getcloudapp.com/jkuZvBLY

## Added tests?


## Added to documentation?


## [optional] What gif best describes this PR or how it makes you feel?

![I promise I'll get you out](https://media.giphy.com/media/xT1R9ZS49VrHph8Lyo/giphy.gif)
